In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [8]:
dataset = load_breast_cancer()
print(dataset.data)
print("------------------------------------------------------------------------------------")
print("Feature names:", dataset.feature_names)
print("Data shape:", dataset.data.shape)
print("Target name:", dataset.target_names)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
------------------------------------------------------------------------------------
Feature names: ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'wor

In [9]:
X = dataset.data
y = dataset.target



In [11]:

# convert data into data frame
df = pd.DataFrame(data=dataset.data,columns=dataset.feature_names)
df


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [12]:
# Displaying Duplicated Values

df.duplicated().sum()



np.int64(0)

In [14]:
# Displaying null values in each columns
df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
dtype: int64

In [16]:
## Training and Testing 
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [17]:
# Initialsing Scaling 

scaler = StandardScaler()

In [18]:
scaled_X_train = scaler.fit_transform(X_train)

In [19]:
scaled_X_test = scaler.transform(X_test)

### Importing Libraries

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier

### For Evaluation

In [28]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [30]:
# Define function for executing the model

def build_model(model,X_train,y_train,X_test,y_test):
    print("model_name : ",model)
    model.fit(X_train,y_train)
    result = model.predict(X_test)
    print("-------------------------------------")
    # Evaluate performance
    print("Accuracy:", accuracy_score(y_test, result))
    print("Classification Report:\n", classification_report(y_test, result))
    print("Confusion Matrix:\n", confusion_matrix(y_test, result)) 
    



In [22]:
## For Storing the Data 

results_list=[]

# Logistic Regression


    Logistic Regression is a linear classification algorithm that models the probability of an instance belonging to a particular class using the sigmoid function.

    It calculates a linear combination of input features and applies the sigmoid to output a probability between 0 and 1.

Suitable for :

    Suitable when relationship between features and target is approximately linear.

    Fast, interpretable, and works well for binary classification or multiclass.

    Performs well when features are scaled and dataset is not highly complex


In [31]:
lr_model = LogisticRegression()
build_model(lr_model,scaled_X_train,y_train,scaled_X_test,y_test)



model_name :  LogisticRegression()
-------------------------------------
Accuracy: 0.9736842105263158
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114

Confusion Matrix:
 [[41  2]
 [ 1 70]]


# Decision Tree Classifier



    Splits data recursively based on feature values that maximize information gain or minimize impurity (e.g. Gini, entropy).

    Creates a tree-like model where each node represents a feature split, and leaves represent class labels.

Suitable for :

    Good for capturing non-linear relationships and feature interactions.

    Easy to interpret and visualize.

    Can overfit on noisy data if not pruned properly.



In [32]:
dtree_model = DecisionTreeClassifier()
build_model(dtree_model,scaled_X_train,y_train,scaled_X_test,y_test)

model_name :  DecisionTreeClassifier()
-------------------------------------
Accuracy: 0.9473684210526315
Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.96      0.96      0.96        71

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114

Confusion Matrix:
 [[40  3]
 [ 3 68]]


# Random Forest Classifier



    An ensemble of decision trees, each trained on random subsets of data and features (bagging + feature randomness).

    Outputs the majority class prediction across all trees.

Suitable for :

    Reduces overfitting compared to single decision trees.

    Handles non-linear relationships, missing values, and high dimensionality effectively.

    Suitable for most classification problems due to its robustness and generalization.




In [34]:
rfc_model = RandomForestClassifier()
build_model(rfc_model,scaled_X_train,y_train,scaled_X_test,y_test)

model_name :  RandomForestClassifier()
-------------------------------------
Accuracy: 0.9649122807017544
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114

Confusion Matrix:
 [[40  3]
 [ 1 70]]


# Support Vector Machine (SVM)


    Finds the optimal hyperplane that maximizes the margin between classes.

    Uses kernels (linear, polynomial, RBF) to handle linear or non-linear data by transforming it into higher dimensions.

Suitable for :

    Effective in high-dimensional spaces and with clear margin separation.

    Good for small to medium-sized datasets where number of features is large.

    Requires careful parameter tuning and scaling of features.





In [36]:
svc_model = SVC()
build_model(svc_model,scaled_X_train,y_train,scaled_X_test,y_test) 



model_name :  SVC()
-------------------------------------
Accuracy: 0.9824561403508771
Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.95      0.98        43
           1       0.97      1.00      0.99        71

    accuracy                           0.98       114
   macro avg       0.99      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114

Confusion Matrix:
 [[41  2]
 [ 0 71]]


# k-Nearest Neighbors (k-NN)



    Instance-based learning (lazy learning) where no explicit model is built.

    Classifies a data point based on the majority label among its k nearest neighbors using a distance metric (e.g. Euclidean).

Suitable for :

    Simple and effective for small datasets with clear clusters.

    Computationally expensive for large datasets as it stores all data points.

    Sensitive to feature scaling and irrelevant features.



In [38]:
knn_model = KNeighborsClassifier(n_neighbors=4)  
build_model(knn_model,scaled_X_train,y_train,scaled_X_test,y_test)

model_name :  KNeighborsClassifier(n_neighbors=4)
-------------------------------------
Accuracy: 0.956140350877193
Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.95      0.94        43
           1       0.97      0.96      0.96        71

    accuracy                           0.96       114
   macro avg       0.95      0.96      0.95       114
weighted avg       0.96      0.96      0.96       114

Confusion Matrix:
 [[41  2]
 [ 3 68]]


C:\Users\MOHAMMED YAZIN N\anaconda3\Lib\site-packages\joblib\externals\loky\backend\context.py:136: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] The system cannot find the file specified
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "C:\Users\MOHAMMED YAZIN N\anaconda3\Lib\site-packages\joblib\externals\loky\backend\context.py", line 257, in _count_physical_cores
    cpu_info = subprocess.run(
        "wmic CPU Get NumberOfCores /Format:csv".split(),
        capture_output=True,
        text=True,
    )
  File "C:\Users\MOHAMMED YAZIN N\anaconda3\Lib\subprocess.py", line 554, in run
    with Popen(*popenargs, **kwargs) as process:
         ~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\MOHAMMED YAZIN N\anaconda3\Lib\subprocess.py", line 1039, in __init__
    self._execute_child(args, executable, preexec_fn, close_fd



## Best Model Performance Report

    Model Name: Support Vector Machine (SVM)
    Algorithm: sklearn.svm.SVC()

  1.  Overall Accuracy Accuracy: 0.982 (98.25%)

  2.  Classification Report

    Macro Average: Precision: 0.99 | Recall: 0.98 | F1-Score: 0.98

    Weighted Average: Precision: 0.98 | Recall: 0.98 | F1-Score: 0.98

    Precision: Very high for both classes, indicating few false positives.
    Recall: Perfect (1.00) for class 1, meaning no false negatives for class 1; slightly lower (0.95) for class 0.
    F1-Score: Overall excellent balance between precision and recall for both classes.

   3. Confusion Matrix

    False Positives (class 0 predicted as 1): 2
    False Negatives (class 1 predicted as 0): 0




## Worst Model Performance Report

Model Name: DecisionTreeClassifier
    Algorithm: sklearn.tree.DecisionTreeClassifier
   
 DecisionTreeClassifier was the least effective among the models tested,despite achieving a respectable accuracy of 94.7%. 

The confusion matrix shows a few misclassifications in both classes — notably, 3 false positives and 3 false negatives. 

While it maintains high precision and recall, its tendency to overfit and limited generalization performance compared to other models resulted in slightly lower overall reliability, making it the weakest performer in this context.